ANN: check existence and type of parameters in format macros

Kobzol opened the pull request:
This PR builds on top of https://github.com/intellij-rust/intellij-rust/pull/3869 and adds additional checks which look for missing or invalid parameters in format macros. After this check exists we can possibly add quick fixes to add missing parameters.

Questions:
- I need to check if a `Ty` implements a given trait. I tried to use `isTraitImplemented` from `ImplLookup`, but I had to make it public and it also doesn't seem to work in tests, I'm probably doing something wrong.
- Should this check be performed every time? Now I only do the check if there are not large errors in the format (for example missing `}`). If there is such a fundamental error, it will probably produce a lot of (false positive) parameter errors, so the parameter check should probably be skipped.

Fixes: https://github.com/intellij-rust/intellij-rust/issues/4089
Umbrella issue: https://github.com/intellij-rust/intellij-rust/issues/2256

vlad20012 commented:
1.1 Use `ImplLookup.canSelect` that is used in `isTraitImplemented` implementation. It has a more complex signature but helps to avoid bugs with generic traits (it's not your case, but anyway).

1.2 The test failed because these traits are implemented for primitives via macros. Impls from macros are work in tests only with `@ExpandMacros(MacroExpansionScope.ALL, "std")` annotation. I'd recommend to use custom types and implement format traits manually instead of primitives.

2. Seems right. And anyway it looks like it's simple to change such behavior in the future.

Kobzol commented:
Thanks, now it works fine. The tests are a bit too verbose to my liking, I need to put `<FORMAT_SPECIFIER>` everywhere. I managed to "turn that off" using `ignoreExtraHighlighting = true`, but then it also ignores errors in the output, which is bad. But it's not a big deal.

vlad20012 commented on src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt:
```diff
@@ -13,29 +13,49 @@ import com.intellij.openapiext.isUnitTestMode
+import org.rust.lang.core.types.type
 import org.rust.lang.utils.parseRustStringCharacters
 
+sealed class FormatArgument(val range: TextRange, val match: MatchResult, val type: String) {
```
MatchResult is not used?

vlad20012 commented:
About verbose tests: we usually create different annotators for highlighting and errors to deal with it, but in this case, it is undesirable because a format pattern would be parsed twice. I found a way to ignore a specific HighlightSeverity in a test, but it looks too hacky to me, so let's leave with these verbose tests for now...

Undin commented:
@vlad20012 Maybe it can be solved by extracting tests with semantic checking into separate class? In new class, we may omit `annotationFixture.registerSeverities(RsColor.values().map(RsColor::testSeverity))` line that seems should fix issue

vlad20012 commented:
@Undin No, it doesn't help. The solution is much hackier :(

Kobzol commented:
Regarding 4: good idea. Two questions:

```rust
#[derive(Debug)]
struct S;

println!("{0} {0:?}", S);
```
- should this highlight `S`? It would be good for consistency, but we maybe want to add a link to the offending parameter in the error.

Also when something like this happens:
```rust
println!("{}");
```
There's no argument to highlight, so should we highlight the closing brace of the `println!` call?

Kobzol commented:
Do we want to supports things like `log::info!`? It's a very common case, so it would be nice. But it would also push the current heuristic for checking if a macro is a formatting macro to its limit. Can we instead somehow expand the macro and see if it's parameters are used in `::core::fmt::Arguments::new_v1` or something like that?

vlad20012 commented:
@Kobzol I'd follow Rustc's messages style. In the case of `println!("{}");` it places annotation to `{}`. In other cases - to an argument.

vlad20012 commented:
Regarding logging... I hope in the future these special cases will not be needed because of stuff like #4640. These macros expand to std's `format_args!` and so eventually should be supported without any additional effort.

Kobzol commented:
> @Kobzol I'd follow Rustc's messages style. In the case of println!("{}"); it places annotation to {}. In other cases - to an argument.

Right, I'm now cross-checking the errors with rustc to make them as similar as possible. So far I have unified `{0}` and `{}` though, Rust has special error messages for the latter, but it seems to me that it would bring unnecessary complexity to the annotator. I'll write as many tests as I can come up with now and we'll see.

Kobzol commented:
Added more tests and refactored the annotator slightly.

Kobzol commented:
Rebased onto master to fix conflict.

Kobzol commented:
@vlad20012 Let me know if there are any other changes that I should make :)

vlad20012 commented:
Looks like this implementation doesn't take into account width/precision arguments. So there are a lot of false-positives in e.g. "liballoc/fmt.rs" (`std::fmt`) examples:
![image](https://user-images.githubusercontent.com/3221931/81185445-4571db00-8fba-11ea-8c0a-c7cdca7b70af.png)

vlad20012 commented on src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt:
```diff
@@ -13,124 +13,123 @@ import com.intellij.openapiext.isUnitTestMode
+        }
     }
+}
+private class NamedParameter(matchInfo: ParameterMatchInfo, val name: String) : FormatParameter(matchInfo)
```
We usually put sealed class children into the parent sealed class. I.e. `FormatParameter.Named` and `FormatParameter.Positional`

vlad20012 commented on src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt:
```diff
@@ -13,124 +13,123 @@ import com.intellij.openapiext.isUnitTestMode
+    val namedParameters: Set<NamedParameter> = parameters.filterIsInstance<NamedParameter>().toSet()
+    val positionalParameters: List<PositionalParameter> = parameters.filterIsInstance<PositionalParameter>().toList()
+
+    val namedArguments: Set<RsFormatMacroArg> = arguments.filter { it.name() != null }.toSet()
```
`Map<String, RsFormatMacroArg>` ?

vlad20012 commented on src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt:
```diff
@@ -13,124 +13,123 @@ import com.intellij.openapiext.isUnitTestMode
+private data class ParsedParameter(val completeMatch: MatchResult,
+                                   val innerContentMatch: MatchResult? = null) {
+    val innerContent = completeMatch.groups[2]
+    val range: IntRange = completeMatch.range
```
Looks like `range` is mapped in all usages (`ctx.toSourceRange(parameter.range)`), so maybe it's possible to map it `toSourceRange` once?

Kobzol replied:
It turned out that once I added the new width/precision specifier parsing, this method became more useful, so I kept it like this.

vlad20012 commented on src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt:
```diff
@@ -140,5 +139,244 @@ class RsFormatMacroAnnotator : AnnotatorBase() {
+    return null
 }
+
+private fun checkArgument(argument: RsFormatMacroArg,
```
formatting

vlad20012 commented on src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt:
```diff
@@ -140,5 +139,244 @@ class RsFormatMacroAnnotator : AnnotatorBase() {
+                val identifier = id.value.toIntOrNull()
+                if (identifier == null) {
+                    NamedParameter(matchInfo, id.value)
+                } else PositionalParameter(matchInfo, identifier)
```
formatting

Kobzol commented:
Wow, I had no idea something like that is even in the `std::fmt` spec :) I added support for it, but here I chose to use a different error message than rustc, because its error is not very nice:
![image](https://user-images.githubusercontent.com/4539057/81273438-2a9d7600-904f-11ea-9fd4-fe55490f556c.png)

I added an error that says `Width specifier must be of type usize`/`Precision specifier must be of type usize`. There is however a slight type inference issue. Rust requires that the type of the width/precision specifier is `usize`.

This works in Rust:
```rust
println!("Hello {:1$}!", "x", 5);
println!("Hello {:1$}!", "x", 5usize);
```
but this sadly does not:
```rust
println!("Hello {:1$}!", "x", 5i32);
```

The problem is that type inference parses `5` as `i32`, so it fails in the plugin. I added a workaround which first checks if the type is `usize` and if not, it still allows integer literals without suffix. It doesn't work for something like `&5` though, which is allowed by Rustc, but that seems like an edge case. I can add support for unwrapping unary expressions (like `&`) or solving constant cases like `5+5`, but it seems like needless complexity at this point, for not a lot gain.

This seems to me like an issue that should be solved directly in Rust, I see no reason not to allow for example `u32` to be used instead of `usize`, I doubt that anyone would ever need more than 2^32 space characters or more than 2^32 places of precision. I'll try to send a PR to Rustc.

I also made error tracking more precise:
![image](https://user-images.githubusercontent.com/4539057/81290534-89231e00-9068-11ea-892f-cee49c2b8718.png)
Now only if there is an implicit `{}` parameter the error will span the whole `{..}` range, otherwise only the specific named/positional parameter will be highlighted.

Kobzol commented:
Hmm, that's weird, build in CI failed, but it's fine locally. I'll try to rebase on master.

vlad20012 commented:
Hmm, there is really an issue with type inference, and it's hard to fix... I don't like false-positives like this:
```rust
let a = 5;
println!("Hello {:1$}!", "x", a);
```
I suggest to just allow i32 (only) along with usize as a width/precision. 

> Hmm, that's weird, build in CI failed, but it's fine locally. I'll try to rebase on master.

Yeah, CI sometimes do rebase before tests run, so try to rebase and fix compilation

Kobzol commented:
Now both `usize` and `i32` are allowed and format trait errors are more precise. I noticed that when I mark the same text span as a format specifier and as an error, it will not highlight the span:
![image](https://user-images.githubusercontent.com/4539057/81437937-36328f00-916c-11ea-8b09-758b2755b776.png)
Is there any workaround for that or is it intended?

vlad20012 commented:
> Is there any workaround for that or is it intended?

Hmm, I'd say it's not intended, but I can't suggest a workaround for now =/ I think we can fix it later.

Still, there are false-positives in "liballoc/fmt.rs":

![image](https://user-images.githubusercontent.com/3221931/81496636-a868b800-92c1-11ea-8048-7b4ee08ce814.png)

Kobzol commented:
Thanks for pointing that out, I missed it. This is a very ugly part of `std::fmt`. The documentation about the precision asterisk is plainly wrong, but sadly it has not been fixed yet (https://github.com/rust-lang/rust/issues/32391). I have implemented it not according to the documentation but according to how Rustc actually behaves (each asterisk precision consumes an unnamed positional argument). It makes the logic convoluted, but hopefully it's the last bit of logic that was missing. There is one remaining false positive in `liballoc/fmt.rs`, but it seems to be related to type inference in doctests, as it works when the same snippet is used in non-documentation code.

vlad20012 commented:
Ok, let's merge it! Please, squash commits that you think require squashing.

vlad20012 commented:
bors r+
